### PR TITLE
executor: fix load data panic if the data is broken at escape character

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31589

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
fix load data will panic in some case
```

### CODE REVIEW

**Review by youjiali1995 — DISMISSED** [review]

**Review by youjiali1995 — COMMENTED** [review]
LGTM

**Comment by jackysp** [comment]
PTAL @lysu

File: executor/load_data.go [hunk 1]
```diff
@@ -461,7 +417,7 @@ func (e *LoadDataInfo) IndexOfTerminator(bs []byte, inQuoter bool) int {
 	atFieldStart := true
 loop:
 	for i := 0; i < len(bs); i++ {
-		if atFieldStart && bs[i] == e.FieldsInfo.Enclosed {
+		if atFieldStart && e.FieldsInfo.Enclosed != byte(0) && bs[i] == e.FieldsInfo.Enclosed {
```
Comment (youjiali1995) [inline executor/load_data.go:420]: Should L426 add this check too?

File: executor/load_data.go [hunk 2]
```diff
@@ -505,67 +461,32 @@ loop:
 // getLine returns a line, curData, the next data start index and a bool value.
```
Comment (youjiali1995) [inline executor/load_data.go:461]: Comments need to be updated.

File: executor/load_data.go [hunk 3]
```diff
@@ -505,67 +461,32 @@ loop:
...
-
-	// terminated symbol in the curData
-	prevData = append(prevData, curData[:nextDataIdx]...)
+	var endIdx int
 	if ignore {
-		endIdx = strings.Index(string(hack.String(prevData[startingLen:])), e.LinesInfo.Terminated)
+		endIdx = strings.Index(string(hack.String(curData[startLen:])), e.LinesInfo.Terminated)
 	} else {
-		endIdx = e.IndexOfTerminator(prevData[startingLen:], inquotor)
+		endIdx = e.IndexOfTerminator(curData[startLen:], false)
```
Comment (youjiali1995) [inline executor/load_data.go:482]: Seems the second parameter `inQuoter` is always false now.

File: executor/load_data.go (hunk 2 above)
Comment (xiongjiwei) [reply executor/load_data.go:461]: comments here no need to change, `getLine` keeps its original logic

Comment (youjiali1995) [inline executor/load_data.go:461]:
```suggestion
// getLine returns a line, curData and a bool value.
```

**Review by youjiali1995 — APPROVED** [review]

**Comment by xiongjiwei** [comment]
/cc @cfzjywxk @zimulala 
PTAL

File: executor/write_test.go [hunk 4]
```diff
@@ -2104,33 +2103,13 @@ func TestLoadDataEscape(t *testing.T) {
 		{nil, []byte("7\trtn0ZbN\n"), []string{"7|" + string([]byte{'r', 't', 'n', '0', 'Z', 'b', 'N'})}, nil, trivialMsg},
 		{nil, []byte("8\trtn0Zb\\N\n"), []string{"8|" + string([]byte{'r', 't', 'n', '0', 'Z', 'b', 'N'})}, nil, trivialMsg},
 		{nil, []byte("9\ttab\\	tab\n"), []string{"9|tab	tab"}, nil, trivialMsg},
+		{[]byte("1\ta string\\"), []byte("\n1\n"), []string{"1|a string\n1"}, nil, trivialMsg},
```
[the remark this answers is not in the document]
Comment (Defined2014) [reply executor/write_test.go:2106]: Please update comments above `// data1 = nil, data2 != nil`

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**